plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -229,83 +228,88 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+                    continue;
+                }
+
+                if (recordScannedFiles) {
+                    // Positional and Equality deletes can only be cleaned up if the whole table has been optimized.
+                    // Equality deletes may apply to many files, and position deletes may be grouped together. This makes it difficult to know if they are obsolete.
+                    List<org.apache.iceberg.DeleteFile> fullyAppliedDeletes = tableHandle.getEnforcedPredicate().isAll() ? wholeFileTask.deletes() : ImmutableList.of();
+                    scannedFiles.add(new DataFileWithDeleteFiles(wholeFileTask.file(), fullyAppliedDeletes));
+                }
+
+                if (fileHasNoDeletions) {
+                    // There were no deletions, so we will produce splits covering the whole file
+                    outputRowsLowerBound = saturatedAdd(outputRowsLowerBound, wholeFileTask.file().recordCount());
+                }
+
+                if (fileHasNoDeletions && noDataColumnsProjected(wholeFileTask)) {
                     fileTasksIterator = List.of(wholeFileTask).iterator();
                 }
                 else {
                     fileTasksIterator = wholeFileTask.split(targetSplitSize).iterator();
                 }
-                fileHasAnyDeletions = false;
                 // In theory, .split() could produce empty iterator, so let's evaluate the outer loop condition again.
                 continue;
             }
-            FileScanTask scanTask = fileTasksIterator.next();
-            fileHasAnyDeletions = fileHasAnyDeletions || !scanTask.deletes().isEmpty();
-            if (scanTask.deletes().isEmpty() &&
-                    maxScannedFileSizeInBytes.isPresent() &&
-                    scanTask.file().fileSizeInBytes() > maxScannedFileSizeInBytes.get()) {
-                continue;
-            }
+            splits.add(toIcebergSplit(fileTasksIterator.next()));
+        }
+        return completedFuture(new ConnectorSplitBatch(splits, isFinished()));
+    }
 
-            if (!pathDomain.includesNullableValue(utf8Slice(scanTask.file().path().toString()))) {
-                continue;
-            }
-            if (!fileModifiedTimeDomain.isAll()) {
-                long fileModifiedTime = getModificationTime(scanTask.file().path().toString());
-                if (!fileModifiedTimeDomain.includesNullableValue(packDateTimeWithZone(fileModifiedTime, UTC_KEY))) {
-                    continue;
-                }
+    private boolean pruneFileScanTask(FileScanTask fileScanTask, boolean fileHasNoDeletions, TupleDomain<IcebergColumnHandle> dynamicFilterPredicate)
+    {
+        if (fileHasNoDeletions &&
+                maxScannedFileSizeInBytes.isPresent() &&
+                fileScanTask.file().fileSizeInBytes() > maxScannedFileSizeInBytes.get()) {
+            return true;
+        }
+
+        if (!pathDomain.includesNullableValue(utf8Slice(fileScanTask.file().path().toString()))) {
```

Comment:
Could avoid the String -> Slice conversion if the domain is all, similar to file modified time below.